### PR TITLE
Allow etcd to store protobuf objects

### PR DESCRIPTION
**Pull request by smarterclayton**
Split storage serialization from client negotiation, and allow API server to take flag controlling serialization.

TODO:
- [x] API server still doesn't start - range allocation object doesn't seem to round trip correctly to etcd
- [ ] Verify that third party resources are ignoring protobuf (add a test)
- [ ] Add integration tests that verify storage is correctly protobuf
- [ ] Add a global default for which storage format to prefer?

### CODE REVIEW

**Comment by smarterclayton** [comment]
@deads2k for the commits that change up storage initialization, @wojtek-t if you want to glance over this.  Still working to clean up ugly bits.

**Comment by smarterclayton** [comment]
@wojtek-t we have some more changes to the client logic to make - we need to initialize a proper codec (easier with this PR) that supports protobuf from the client, instead of using LegacyCodec which is JSON only.  Right now we're sending JSON with a mime-type protobuf.  We'll need that to enable test-cmd to run using proto:

```
KUBE_TEST_API_TYPE="application/vnd.kubernetes.protobuf" KUBE_TEST_API_STORAGE_TYPE="application/vnd.kubernetes.protobuf" hack/test-cmd.sh 
```

File: pkg/genericapiserver/storage_factory.go [hunk 1]
```diff
@@ -219,3 +226,25 @@ func (s *DefaultStorageFactory) Backends() []string {
 	}
 	return backends.List()
 }
+
+// NewStorageCodec assembles a storage codec for the provided storage media type, the provided serializer, and the requested
+// storage and memory versions.
+func NewStorageCodec(storageMediaType string, ns runtime.StorageSerializer, storageVersion, memoryVersion unversioned.GroupVersion) (runtime.Codec, error) {
+	s, ok := ns.SerializerForMediaType(storageMediaType, nil)
+	if !ok {
+		return nil, fmt.Errorf("unable to find serializer for JSON")
```
Comment (deads2k) [inline pkg/genericapiserver/storage_factory.go]: not necessarily json

Comment (smarterclayton) [inline pkg/genericapiserver/storage_factory.go]: Fixed in a later commit (spotted that as well)

File: pkg/apiserver/apiserver.go [hunk 2]
```diff
@@ -277,9 +274,13 @@ type StripVersionNegotiatedSerializer struct {
 	runtime.NegotiatedSerializer
 }
 
-func (n StripVersionNegotiatedSerializer) EncoderForVersion(serializer runtime.Serializer, gv unversioned.GroupVersion) runtime.Encoder {
-	encoder := n.NegotiatedSerializer.EncoderForVersion(serializer, gv)
-	return stripVersionEncoder{encoder, serializer}
+func (n StripVersionNegotiatedSerializer) EncoderForVersion(encoder runtime.Encoder, gv unversioned.GroupVersion) runtime.Encoder {
+	serializer, ok := encoder.(runtime.Serializer)
```
Comment (deads2k) [inline pkg/apiserver/apiserver.go]: if you want a serializer, why did you change the sig?

Comment (deads2k) [inline pkg/apiserver/apiserver.go]: if you need an interface that implements both, create that interface and use it.

File: pkg/apiserver/apiserver_test.go [hunk 3]
```diff
@@ -1234,8 +1242,9 @@ func TestMetadata(t *testing.T) {
 		}
 	}
 	if matches["text/plain,application/json,application/yaml,application/vnd.kubernetes.protobuf"] == 0 ||
+		matches["application/json,application/json;stream=watch,application/vnd.kubernetes.protobuf,application/vnd.kubernetes.protobuf;stream=watch"] == 0 ||
```
Comment (deads2k) [inline pkg/apiserver/apiserver_test.go]: yeouch.

File: pkg/apiserver/apiserver.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/apiserver/apiserver.go]: The overall sig changed, this is a hack because there isn't enough local
info.

File: pkg/apiserver/apiserver_test.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/apiserver/apiserver_test.go]: Tempted to refactor this but it's kind of ugly in general to calculate this
any other way.

File: pkg/runtime/interfaces.go [hunk 4]
```diff
@@ -80,13 +80,81 @@ type ParameterCodec interface {
...
+	// DecoderForVersion returns a decoder that ensures objects being read by the provided
+	// serializer are in the provided group version by default.
+	// TODO: take multiple group versions
+	DecoderToVersion(serializer Decoder, gv unversioned.GroupVersion) Decoder
+}
+
+// StorageSerializer is an interface used for obtaining encoders, decoders, and serializers
+// that can read and write data at rest. This would commonly be used by client tools that must
+// read files, or server side storage interfaces that persist restful objects.
+type StorageSerializer interface {
```
Comment (deads2k) [inline pkg/runtime/interfaces.go]: Seems like this should have `SupportedMediaTypes` too.

**Comment by deads2k** [comment]
the storage config bits lgtm

File: pkg/runtime/interfaces.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/runtime/interfaces.go]: Probably at some point, did not expose it yet until we need it.

File: docs/admin/kube-apiserver.md [hunk 5]
```diff
@@ -109,6 +109,7 @@ kube-apiserver
       --service-node-port-range=: A port range to reserve for services with NodePort visibility.  Example: '30000-32767'.  Inclusive at both ends of the range.
       --ssh-keyfile="": If non-empty, use secure SSH proxy to the nodes, using this user keyfile
       --ssh-user="": If non-empty, use secure SSH proxy to the nodes, using this user name
+      --storage-media-type="application/json": The media type to use to store objects in storage. Defaults to application/json. Some resources may override this value.
```
Comment (lavalamp) [inline docs/admin/kube-apiserver.md]: Hm, do we want this to maybe be a default with a separate override list?

Comment (lavalamp) [inline docs/admin/kube-apiserver.md]: i.e., change to `--storage-media-type-default`. Introduce the override list when it becomes necessary.

File: hack/test-cmd.sh [hunk 6]
```diff
@@ -148,6 +148,9 @@ HOME="${KUBE_TEMP}"
 kube::log::status "Running kubectl with no options"
 "${KUBE_OUTPUT_HOSTBIN}/kubectl"
 
+# Only run kubelet on platforms it supports
+if [[ "$(go env GOHOSTOS)" == "linux" ]]; then
```
Comment (lavalamp) [inline hack/test-cmd.sh]: unrelated to this change?

File: hack/test-cmd.sh [hunk 7]
```diff
@@ -202,11 +208,31 @@ kube::util::wait_for_url "http://127.0.0.1:${API_PORT}/healthz" "apiserver"
...
 "${KUBE_OUTPUT_HOSTBIN}/kube-controller-manager" \
   --port="${CTLRMGR_PORT}" \
+  --kube-api-content-type="${KUBE_TEST_API_TYPE-}" \
   --master="127.0.0.1:${API_PORT}" 1>&2 &
 CTLRMGR_PID=$!
 
 kube::util::wait_for_url "http://127.0.0.1:${CTLRMGR_PORT}/healthz" "controller-manager"
-kube::util::wait_for_url "http://127.0.0.1:${API_PORT}/api/v1/nodes/127.0.0.1" "apiserver(nodes)"
+
+if [[ "$(go env GOHOSTOS)" == "linux" ]]; then
```
Comment (lavalamp) [inline hack/test-cmd.sh]: also unrelated?

File: docs/admin/kube-apiserver.md (hunk 5 above)
Comment (smarterclayton) [inline docs/admin/kube-apiserver.md]: Sorry, the description I meant was "third party resources will never use
this".  "Some resource types may ignore this setting." would be a more
accurate message.

File: hack/test-cmd.sh (hunk 7 above)
Comment (smarterclayton) [inline hack/test-cmd.sh]: It's been split to its own PR

**Comment by smarterclayton** [comment]
Rebased and addressed the initial comments.  The first commit is in the merge queue and when that merges this will be rebased without it.

**Comment by smarterclayton** [comment]
This is ready for review (there are a few other tests I'm looking to add), but is based on top of the etcd3 work (so binary encoding is only necessary on etcd2 stacks).

**Comment by smarterclayton** [comment]
The first commit is in another PR in the merge queue (stripping protobuf imports).

File: pkg/genericapiserver/storage_factory.go [hunk 8]
```diff
@@ -217,3 +233,39 @@ func (s *DefaultStorageFactory) Backends() []string {
...
+		return nil, fmt.Errorf("%q is not a valid mime-type", storageMediaType)
+	}
+	serializer, ok := ns.SerializerForMediaType(mediaType, options)
+	if !ok {
+		return nil, fmt.Errorf("unable to find serializer for %q", storageMediaType)
+	}
+
+	s := serializer.Serializer
+
+	// etcd2 only supports string data - we must wrap any result before returning
```
Comment (hongchaodeng) [inline pkg/genericapiserver/storage_factory.go]: I'm confused here.
In etcd, string are just immutable representation of []byte.
Why do we need to base64-encode it again?

/cc @xiang90

Comment (xiang90) [inline pkg/genericapiserver/storage_factory.go]: for etcd2 yes. etcd2 only accepts json. json string cannot contain random bytes, you have to base64 it.

**Comment by wojtek-t** [comment]
@smarterclayton - can you please rebase

@lavalamp - can you please review, we should prioritize it?

File: pkg/runtime/serializer/recognizer/recognizer.go [hunk 9]
```diff
@@ -27,18 +27,38 @@ import (
 
 type RecognizingDecoder interface {
 	runtime.Decoder
+	// AccurateRecognizer should be true if this decoder can unambigously determine
+	// whether the content is known to it. Allows RecognizingDecoders to adjust certainty.
+	AccurateRecognizer() bool
+	// RecognizesData should return true if the input provided in the provided reader
+	// belongs to this decoder, or an error if the data could not be read or is ambiguous.
```
Comment (lavalamp) [inline pkg/runtime/serializer/recognizer/recognizer.go]: document: is it allowed to consume data from the Reader?

File: pkg/runtime/serializer/recognizer/recognizer.go [hunk 10]
```diff
@@ -27,18 +27,38 @@ import (
...
+//
+// 1. The decoder implements RecognizingDecoder and returns true for AccurateRecognizer
+// 2. The decoder implements RecognizingDecoder and returns false for AccurateRecognizer
+// 3. All other decoders
+//
+// The order passed to the constructor is preserved within those priorities. Only the
+// last error encountered will be returned.
 func NewDecoder(decoders ...runtime.Decoder) runtime.Decoder {
 	recognizing, blind := []RecognizingDecoder{}, []runtime.Decoder{}
+	var ambiguous []runtime.Decoder
```
Comment (lavalamp) [inline pkg/runtime/serializer/recognizer/recognizer.go]: nit: combine with above line or split into three var lines?

File: pkg/runtime/serializer/recognizer/recognizer.go [hunk 11]
```diff
@@ -27,18 +27,38 @@ import (
...
+			if r.AccurateRecognizer() {
+				recognizing = append(recognizing, r)
+			} else {
+				ambiguous = append(ambiguous, r)
+			}
 		} else {
 			blind = append(blind, d)
 		}
 	}
+	blind = append(ambiguous, blind...)
```
Comment (lavalamp) [inline pkg/runtime/serializer/recognizer/recognizer.go]: Shouldn't you append ambiguous with the recognizers instead?

Comment (smarterclayton) [inline pkg/runtime/serializer/recognizer/recognizer.go]: No, but I need to clarify why as a godoc - ambiguous can't guarantee that
no means no, so if we put it above other recognizers we get false
negatives.  The ordering of the input decoders is important, and the
guarantee that certainty comes first

File: pkg/runtime/serializer/recognizer/recognizer.go [hunk 12]
```diff
@@ -27,18 +27,38 @@ import (
 
 type RecognizingDecoder interface {
 	runtime.Decoder
+	// AccurateRecognizer should be true if this decoder can unambigously determine
+	// whether the content is known to it. Allows RecognizingDecoders to adjust certainty.
```
Comment (lavalamp) [inline pkg/runtime/serializer/recognizer/recognizer.go]: Please clarify if this means it allows false positives, false negatives, or both.

File: pkg/runtime/serializer/recognizer/recognizer.go [hunk 13]
```diff
@@ -27,18 +27,38 @@ import (
...
+	AccurateRecognizer() bool
+	// RecognizesData should return true if the input provided in the provided reader
+	// belongs to this decoder, or an error if the data could not be read or is ambiguous.
 	RecognizesData(peek io.Reader) (bool, error)
```
Comment (lavalamp) [inline pkg/runtime/serializer/recognizer/recognizer.go]: Maybe just consider changing this interface to something like:
* definitely yes I handle this data and no one else could
* I might be able to handle this, but I'm not confident
* I definitely can't handle this data.

Comment (smarterclayton) [inline pkg/runtime/serializer/recognizer/recognizer.go]: Changed it, slightly more complex now.

**Comment by smarterclayton** [comment]
@wojtek-t see changes to NegotiatedSerializerWrapper - should be equivalent, but closer to the new signature.

**Comment by wojtek-t** [comment]
As discussed offline with @lavalamp I'm taking reassigning this for review to myself. Taking a deeper look now.

File: hack/test-cmd.sh [hunk 14]
```diff
@@ -202,6 +203,7 @@ kube::util::wait_for_url "http://127.0.0.1:${API_PORT}/healthz" "apiserver"
 kube::log::status "Starting controller-manager"
 "${KUBE_OUTPUT_HOSTBIN}/kube-controller-manager" \
   --port="${CTLRMGR_PORT}" \
+  --kube-api-content-type="${KUBE_TEST_API_TYPE-}" \
```
Comment (wojtek-t) [inline hack/test-cmd.sh]: nit: we should probably start all other components in the same mode too (scheduler, kubelet, kube-proxy). But we can do it in a separate PR probably.

File: pkg/api/unversioned/generated.proto [hunk 15]
```diff
@@ -21,6 +21,7 @@ syntax = 'proto2';
 
 package k8s.io.kubernetes.pkg.api.unversioned;
 
+import "k8s.io/kubernetes/pkg/runtime/generated.proto";
```
Comment (wojtek-t) [inline pkg/api/unversioned/generated.proto]: Where does those lines come from? You are not touching generator in this PR.

File: pkg/apiserver/negotiate_test.go [hunk 16]
```diff
@@ -41,27 +41,34 @@ func (n *fakeNegotiater) SupportedStreamingMediaTypes() []string {
...
 		n.streamOptions = options
 	}
-	return n.streamSerializer, n.framer, mediaType, n.streamSerializer != nil
+	return runtime.StreamSerializerInfo{
+		SerializerInfo: runtime.SerializerInfo{
+			Serializer:    n.serializer,
+			MediaType:     mediaType,
+			EncodesAsText: true,
+		},
+		Framer: n.framer,
```
Comment (wojtek-t) [inline pkg/apiserver/negotiate_test.go]: Shouldn't we set Embedded too here?

File: pkg/runtime/codec.go [hunk 17]
```diff
@@ -169,3 +170,27 @@ func (c *parameterCodec) EncodeParameters(obj Object, to unversioned.GroupVersio
...
+}
+
+func NewBase64Serializer(s Serializer) Serializer {
+	return &base64Serializer{s}
+}
+
+func (s base64Serializer) EncodeToStream(obj Object, stream io.Writer, overrides ...unversioned.GroupVersion) error {
+	e := base64.NewEncoder(base64.StdEncoding, stream)
+	err := s.Serializer.EncodeToStream(obj, e, overrides...)
+	e.Close()
```
Comment (wojtek-t) [inline pkg/runtime/codec.go]: hmm - do we really want to close it?
That would mean that we cannot encode more than one object to the given stream, which doesn't sound good to me.

File: pkg/runtime/serializer/protobuf/protobuf.go [hunk 18]
```diff
@@ -419,12 +419,6 @@ func (s *RawSerializer) EncodeToStream(obj runtime.Object, w io.Writer, override
 	}
 }
 
-// RecognizesData implements the RecognizingDecoder interface - objects encoded with this serializer
-// have no innate identifying information and so cannot be recognized.
-func (s *RawSerializer) RecognizesData(peek io.Reader) (bool, error) {
-	return false, nil
-}
-
```
Comment (wojtek-t) [inline pkg/runtime/serializer/protobuf/protobuf.go]: Don't we need it?

File: pkg/runtime/serializer/recognizer/recognizer.go [hunk 19]
```diff
@@ -27,50 +28,110 @@ import (
...
+				return out, actual, nil
+			}
+		default:
+			// if this is a non-RecognizingDecoder
+			out, actual, err := r.Decode(data, gvk, into)
+			if err != nil {
+				lastErr = err
+				continue
+			}
+			return out, actual, nil
```
Comment (wojtek-t) [inline pkg/runtime/serializer/recognizer/recognizer.go]: Both branches in this switch seem to be the same.
Wouldn't it be easier to:
1. add 
  default: skipped = append(skipped, t)
  in the previous for loop (the one trying only recognizing decoders)
2. here simply iterate over skilled:
  for _, s := range skipped {
    out, actual, err := s.Decode(data, gvk, into)
   ...

?

File: test/integration/framework/master_utils.go [hunk 20]
```diff
@@ -155,24 +155,29 @@ func NewMasterConfig() *master.Config {
...
 
-	negotiatedSerializer := NewSingleContentTypeSerializer(api.Scheme, testapi.Default.Codec(), "application/json")
+	negotiatedSerializer := NewSingleContentTypeSerializer(api.Scheme, testapi.Default.Codec(), runtime.ContentTypeJSON)
 
-	storageFactory := genericapiserver.NewDefaultStorageFactory(config, negotiatedSerializer, genericapiserver.NewDefaultResourceEncodingConfig(), master.DefaultAPIResourceConfigSource())
+	storageFactory := genericapiserver.NewDefaultStorageFactory(config, runtime.ContentTypeJSON, negotiatedSerializer, genericapiserver.NewDefaultResourceEncodingConfig(), master.DefaultAPIResourceConfigSource())
 	storageFactory.SetSerializer(
 		unversioned.GroupResource{Group: api.GroupName, Resource: genericapiserver.AllResources},
-		NewSingleContentTypeSerializer(api.Scheme, testapi.Default.Codec(), "application/json"))
+		"",
```
Comment (wojtek-t) [inline test/integration/framework/master_utils.go]: Just question: what does empty media type mean?

**Comment by wojtek-t** [comment]
@smarterclayton - I added some minor comments. Other than that LGTM

File: pkg/genericapiserver/storage_factory.go [hunk 21]
```diff
@@ -217,3 +233,39 @@ func (s *DefaultStorageFactory) Backends() []string {
...
+	// TODO: storagebackend should return a boolean indicating whether it supports binary data
+	if !serializer.EncodesAsText && (config.Type == storagebackend.StorageTypeUnset || config.Type == storagebackend.StorageTypeETCD2) {
+		glog.V(4).Infof("Wrapping the underlying binary storage serializer with a base64 encoding for etcd2")
+		s = runtime.NewBase64Serializer(s)
+	}
+
+	ds := recognizer.NewDecoder(s, ns.UniversalDeserializer())
```
Comment (wojtek-t) [inline pkg/genericapiserver/storage_factory.go]: question: does it mean that we will have migration to protobufs in etcd for free?

File: pkg/api/unversioned/generated.proto (hunk 15 above)
Comment (smarterclayton) [inline pkg/api/unversioned/generated.proto]: We're not filtering imports in the protobuf, so if we touch a type that
requires it it'll get added.

On Thu, May 5, 2016 at 2:29 AM, Wojciech Tyczynski <notifications@github.com

File: pkg/apiserver/negotiate_test.go (hunk 16 above)
Comment (smarterclayton) [inline pkg/apiserver/negotiate_test.go]: None of the tests are currently using it - can be added, but it wouldn't do anything

File: pkg/runtime/codec.go (hunk 17 above)
Comment (smarterclayton) [inline pkg/runtime/codec.go]: We're closing the base64 encoder, which is required to flush it.  It
doesn't close the stream.

On Thu, May 5, 2016 at 3:25 AM, Wojciech Tyczynski <notifications@github.com

File: pkg/runtime/serializer/protobuf/protobuf.go (hunk 18 above)
Comment (smarterclayton) [inline pkg/runtime/serializer/protobuf/protobuf.go]: No longer required to implement the interface - raw serializers cannot
recognize data (no magic bytes to look at)

On Thu, May 5, 2016 at 3:33 AM, Wojciech Tyczynski <notifications@github.com

File: test/integration/framework/master_utils.go (hunk 20 above)
Comment (smarterclayton) [inline test/integration/framework/master_utils.go]: I'll set it to json, there's no reason not to here.

File: pkg/genericapiserver/storage_factory.go (hunk 21 above)
Comment (smarterclayton) [inline pkg/genericapiserver/storage_factory.go]: Yes, once the storage type is set registries will start replacing values
that are written with proto.  So the same as a version upgrade - read /
write all values.

On Thu, May 5, 2016 at 4:37 AM, Wojciech Tyczynski <notifications@github.com

File: pkg/runtime/codec.go (hunk 17 above)
Comment (wojtek-t) [inline pkg/runtime/codec.go]: Got it - thanks for explanation

File: pkg/api/unversioned/generated.proto (hunk 15 above)
Comment (wojtek-t) [inline pkg/api/unversioned/generated.proto]: Hmm - I don't understand.
The rest of the file didn't change - so why did the imports change?

File: pkg/genericapiserver/storage_factory.go (hunk 21 above)
Comment (wojtek-t) [inline pkg/genericapiserver/storage_factory.go]: OK - that's how I understood it. Thanks!

File: pkg/api/unversioned/generated.proto (hunk 15 above)
Comment (smarterclayton) [inline pkg/api/unversioned/generated.proto]: Not sure - could be an input dependency order.  I'll open an issue to have
proto only generate imports it uses.

Comment (smarterclayton) [inline pkg/api/unversioned/generated.proto]: Spawned https://github.com/kubernetes/kubernetes/issues/25200

File: test/integration/framework/master_utils.go (hunk 20 above)
Comment (smarterclayton) [inline test/integration/framework/master_utils.go]: Actually this is the override type - that bypasses the behavior of the
serializer itself.  If empty, no override.

On Thu, May 5, 2016 at 3:47 AM, Wojciech Tyczynski <notifications@github.com

Comment (smarterclayton) [inline test/integration/framework/master_utils.go]: Updated with the cleaner recognizer logic.

File: pkg/runtime/serializer/recognizer/recognizer.go [hunk 22]
```diff
@@ -27,51 +28,93 @@ import (
...
+			ok, unknown, err := t.RecognizesData(buf)
+			if err != nil {
+				lastErr = err
+				continue
+			}
+			if unknown || !ok {
+				skipped = append(skipped, t)
+				continue
+			}
+			return r.Decode(data, gvk, into)
```
Comment (wojtek-t) [inline pkg/runtime/serializer/recognizer/recognizer.go]: don't we need:

default:
  skipped = append(skipped, t)

?

**Comment by wojtek-t** [comment]
One very minor comment - feel free to self-apply lgtm label after answering.

**Comment by smarterclayton** [comment]
Rebased, applying

**Comment by wojtek-t** [comment]
thanks @smarterclayton - LGTM

File: cmd/kube-apiserver/app/options/options.go [hunk 23]
```diff
@@ -44,6 +44,7 @@ type APIServer struct {
 	AuthorizationMode          string
 	AuthorizationConfig        apiserver.AuthorizationConfig
 	BasicAuthFile              string
+	DefaultStorageMediaType    string
```
Comment (nikhiljindal) [inline cmd/kube-apiserver/app/options/options.go]: This should go in genericapiserver rather than duplicating in kubernetes-apiserver and federation-apiserver?

File: federation/cmd/federated-apiserver/app/options/options.go [hunk 24]
```diff
@@ -76,6 +77,7 @@ func NewAPIServer() *APIServer {
 		ServerRunOptions:        genericapiserver.NewServerRunOptions(),
 		AdmissionControl:        "AlwaysAdmit",
 		AuthorizationMode:       "AlwaysAllow",
+		DefaultStorageMediaType: "application/json",
```
Comment (nikhiljindal) [inline federation/cmd/federated-apiserver/app/options/options.go]: Use runtime.ContentTypeJSON ?

File: cmd/kube-apiserver/app/options/options.go (hunk 23 above)
Comment (smarterclayton) [inline cmd/kube-apiserver/app/options/options.go]: Will do this in a follow up given priority.  There are other follow ups and @wojtek-t and I need to be enabling this now.

**Comment by smarterclayton** [comment]
@nikhiljindal will open a PR in a few that has your follow up

**Comment by wojtek-t** [comment]
Ref #8132
